skyportal/handlers/api/source.py

```diff
@@ -725,6 +751,78 @@ def get_sources(
+    ):
+        if annotations_filter is not None:
+            for ann_filt in annotations_filter:
+                ann_split = ann_filt.split(":")
+                if not (len(ann_split) == 1 or len(ann_split) == 3):
+                    raise ValueError(
+                        "Invalid annotationsFilter value -- annotation filter must have 1 or 3 values"
+                    )
+                name = ann_split[0].strip()
+
+                annotations_query = create_annotations_query(
+                    session,
+                    annotations_filter_origin=annotations_filter_origin,
+                    annotations_filter_before=annotations_filter_before,
+                    annotations_filter_after=annotations_filter_after,
+                )
+
+                if len(ann_split) == 3:
+                    value = ann_split[1].strip()
+                    try:
+                        value = float(value)
+                    except ValueError as e:
+                        raise ValueError(f"Invalid annotation filter value: {e}")
+                    op = ann_split[2].strip()
+                    op_options = ["lt", "le", "eq", "ne", "ge", "gt"]
+                    if op not in op_options:
+                        raise ValueError(f"Invalid operator: {op}")
+
+                    if op == "lt":
+                        comp_function = operator.lt
+                    elif op == "le":
+                        comp_function = operator.le
+                    elif op == "eq":
+                        comp_function = operator.eq
+                    elif op == "ne":
+                        comp_function = operator.ne
+                    elif op == "ge":
+                        comp_function = operator.ge
+                    elif op == "gt":
+                        comp_function = operator.gt
+
+                    annotations_query = annotations_query.where(
+                        comp_function(Annotation.data[name], cast(value, JSONB))
+                    )
+                else:
+                    annotations_query = annotations_query.where(
+                        Annotation.data[name].astext.is_not(None)
+                    )
+
+                annotations_subquery = annotations_query.subquery()
+                obj_query = obj_query.join(
+                    annotations_subquery,
+                    Obj.id == annotations_subquery.c.obj_id,
+                )
+        else:
+            annotations_query = create_annotations_query(
+                session,
+                annotations_filter_origin=annotations_filter_origin,
+                annotations_filter_before=annotations_filter_before,
+                annotations_filter_after=annotations_filter_after,
+            )
+            annotations_subquery = annotations_query.subquery()
+            obj_query = obj_query.join(
+                annotations_subquery,
+                Obj.id == annotations_subquery.c.obj_id,
+            )
```

Comment:
Can this be outside the `if annotations_filter is not None:`?

Reply:
Unfortunately no because that is a loop inside of there (you can have multiple filters on those annotations).